### PR TITLE
Don't create rendering device or parse glsl shader in headless mode

### LINKED ISSUE #94734: Can't import GLSL shaders in headless mode
### Tested versions

4.3.beta.mono.custom_build.ba7c5b0c8

### System information

Godot v4.3.beta.mono (ba7c5b0c8) - Pop!_OS 22.04 LTS - Wayland - Vulkan (Forward+) - dedicated Intel(R) Arc(tm) A750 Graphics (DG2) - AMD Ryzen 7 5800X 8-Core Processor (16 Threads)

### Issue description

Godot currently doesn't support importing glsl shaders in a headless context.  This creates a problem for continuous integration / pipelines.  @clayjohn worded it perfectly in the rocket chat so I'll just quote it here:

> This is a very valuable workflow when you are targeting multiple platforms and becomes super important when working with a team. Basically everyone on the team can work on the platform of their choice, then the CI makes an export build for all target platforms and then you can test release builds on all platforms

> Since CI doesn't have a GPU, you have to be able to import all resources, then export the game without access to a GPU. Ultimately, the game will be played on a device with a GPU, but the import/export process happens without one

This appears to be the PR that disabled glsl imports in headless mode: https://github.com/godotengine/godot/pull/92539

I tried simply reverting that PR and interestingly, it does seem to import the glsl shader.  However, it crashes in the exported project when that shader is loaded if it wasn't previously loaded in a non-headless context.

### Steps to reproduce

1. Clone the proton scatter repo or use the zipped mrp (which is also just the proton scatter repo): https://github.com/HungryProton/scatter.git
2. Open in godot
3. Notice that the trees are loading in the showcase scene.
4. Create an export preset
5. Delete the ``.godot`` folder in the project directory so there aren't any cached imports.
6. Run the command: ``godot --headless --path path/to/project --import``
7. Notice the errors regarding failed glsl imports.
8. Run the command: ``godot --headless --path path/to/project --export-debug PresetName path/to/export/game.exe``
9. Notice the errors regarding glsl
10. Start the exported game and notice that the trees no longer load

### Minimal reproduction project (MRP)

[scatter.zip](https://github.com/user-attachments/files/16377132/scatter.zip)

https://github.com/HungryProton/scatter.git

### LINKED ISSUE #98246: Headless import without gpu crashes
### Tested versions

Reproduceable in 4.4.dev[92e51fca7247c932f95a1662aefc28aca96e8de6]

### System information

Linux Pop OS 22.04

### Issue description

**NOTE**

This is different from this other similar issue I put in and also have a PR for: https://github.com/godotengine/godot/issues/98062

**Description**

When importing with the ``--headless`` argument on an actual headless server without a gpu, godot crashes with:

```
ERROR: Condition "!tasks.has(p_task)" is true. Returning: canceled
   at: task_step (editor/progress_dialog.cpp:203)
ERROR: Condition "!tasks.has(p_task)" is true.
   at: end_task (editor/progress_dialog.cpp:226)
WARNING: UID duplicate detected between res://submodules/BlockyBallOT-Art/BlockBall/Skin_Materials/Skin_Magma_RedRock.tres and res://material_maker/BlockyGuy/Skin_Magma/Skin_Magma.material.tres.
     at: _process_file_system (editor/editor_file_system.cpp:1272)
ERROR: Cannot import custom .glsl shaders when running in headless mode.
   at: import (editor/import/resource_importer_shader_file.cpp:95)
ERROR: Error importing 'res://addons/proton_scatter/src/modifiers/compute_shaders/compute_relax.glsl'.
   at: _reimport_file (editor/editor_file_system.cpp:2884)
ERROR: Required extension VK_KHR_surface not found.
   at: _initialize_instance_extensions (drivers/vulkan/rendering_context_driver_vulkan.cpp:487)
ERROR: Condition "err != OK" is true. Returning: err
   at: initialize (drivers/vulkan/rendering_context_driver_vulkan.cpp:928)
ERROR: FATAL: Index p_index = 0 is out of bounds (count = 0).
   at: operator[] (./core/templates/local_vector.h:177)
Error: nu::shell::coredump_error

  × program coredump error
     ╭─[/home/vsts/work/1/s/submodules/godot-src/godot/godot.nu:142:5]
 141 │     print $"Running godot command: ($config.godot_bin) ($rest | str join ' ')"
 142 │     run-external $config.godot_bin ...$rest
     ·     ──────┬─────
     ·           ╰── to nushell `Value`invalid entry, dup11
 143 │ }
     ╰────

```

^
My pipeline executes within a nushell script.  The nushell script was simply catching the coredump / crash and displaying it.  Just to clarify on the last error.

Upon investigating the code, some changes went through that are causing godot to attempt loading gpu libs even tho the ``--headless`` arg was passed.  I made some slight changes in my fork to fix this and will have a PR in shortly.

### Steps to reproduce

1. Make sure you are running from within a headless server without a gpu.  Easily accomplished with a Docker container or Podman container.
2. Import a godot project with the --headless argument.
3. If that doesn't work, try ensuring that a glsl shader is in the project as well.
4. See that Godot crashes

### Minimal reproduction project (MRP)

N/A

### PR DESCRIPTION
**Pull request by TCROC**
This PR resolves: https://github.com/godotengine/godot/issues/98246

It appears that some recent code changes were made that parse glsl shaders in new locations as well as some logic surrounding ``DisplayServer::can_create_rendering_device``.  Both of which introduced problems when running on a machine that does not have a gpu.  This may not be the correct solution so please leave your comments and / or open a different pr with the proper fix.  This at least fixed it for me though.  Without further ado, here is the explanation for my changes:

```cpp
bool DisplayServer::can_create_rendering_device() {
	if (get_singleton()->get_name() == "headless") {
		return false;
	}
```

^ We can't create a DisplayServer in headless mode... at least I don't think we can.  I'm mostly going off my intuitive understanding of the word ``Display`` and regarding the fact that there is no display on a headless server without a gpu.  But that is grammatical rather than technical reasoning which I know does not necessarily apply in the world of software.

```cpp
void BetsyCompressor::_init() {
	if (!DisplayServer::can_create_rendering_device()) {
		return;
	}
```

^ Don't try to initalize a new BetsyCompressor in headless mode.  I found that for whatever reason, BetsyCompressor was trying to hook into the gpu on my device and crashing if it couldn't find one.  This may or may not be the correct solution.

```
Error RDShaderFile::parse_versions_from_text(const String &p_text, const String p_defines, OpenIncludeFunction p_include_func, void *p_include_func_userdata) {
	ERR_FAIL_COND_V_EDMSG((DisplayServer::get_singleton()->get_name() == "headless"), ERR_UNAVAILABLE, "Cannot import custom .glsl shaders when running in headless mode.");
	ERR_FAIL_NULL_V(RenderingDevice::get_singleton(), ERR_UNAVAILABLE);
```

^ This is really the only change that I'm fairly confident in.  I actually copied this ``ERROR_FAIL`` from a different area of code that prevents glsl imports in headless mode

https://github.com/godotengine/godot/blob/04692d83cb8f61002f18ea1d954df8c558ee84f7/editor/import/resource_importer_shader_file.cpp#L95

And if you notice, that is actually preventing ``shader_file->parse_versions_from_text`` from being called on line 106: https://github.com/godotengine/godot/blob/04692d83cb8f61002f18ea1d954df8c558ee84f7/editor/import/resource_importer_shader_file.cpp#L106C8-L106C44

This is introduces a crash.  So we might as well copy that check to the root cause method as well for stability purposes.  Which is what I did in my change :)

**NOTE:**

The change I said I was "confident in" will no longer be necessary when this issue gets resolved: https://github.com/godotengine/godot/issues/94734

Which also happens to be an issue I opened 😎.  But until then, this change is necessary to prevent crashes and increase stability.

I look forward to hearing your guys thoughts!  As always, thanks for the awesome engine! :)

### CODE REVIEW

File: servers/rendering/rendering_device_binds.cpp [hunk 1]
```diff
@@ -31,6 +31,7 @@
 #include "rendering_device_binds.h"
 
 Error RDShaderFile::parse_versions_from_text(const String &p_text, const String p_defines, OpenIncludeFunction p_include_func, void *p_include_func_userdata) {
+	ERR_FAIL_COND_V_EDMSG((DisplayServer::get_singleton()->get_name() == "headless"), ERR_UNAVAILABLE, "Cannot import custom .glsl shaders when running in headless mode.");
```
Comment (clayjohn) [inline servers/rendering/rendering_device_binds.cpp:34]: This should already fail on the next line. The fact that it isn't highlights that `get_singleton()` is returning a valid pointer when it shouldn't. Therefore the solution is going to be ensuring that the singleton isn't initialized when RD isn't available

Comment (TCROC) [reply servers/rendering/rendering_device_binds.cpp:34]: You are correct that it does indeed fail when get_singleton is null.  I believe my 

```cpp
void BetsyCompressor::_init() {
	if (!DisplayServer::can_create_rendering_device()) {
		return;
	}
```

and

```
bool DisplayServer::can_create_rendering_device() {
	if (get_singleton()->get_name() == "headless") {
		return false;
	}
```

Fixes the issue of the singleton being created.

But then on top of that, I just thought this error message was more helpful than the null error message regarding the singleton.  If you prefer, I can remove the change you commented on and I believe it will still work.  It will just throw "null get_singleton" error messages instead of the "Cannot import custom .glsl shaders when running in headless mode." error messages.  Whichever you prefer is fine with me.

TL; DR;

You are correct.  My other 2 changes fix the issue of that singleton being created.  If you would like, I can remove this change.

Comment (clayjohn) [reply servers/rendering/rendering_device_binds.cpp:34]: It is a much better error message, but I would prefer to improve the error message for the old error instead of doing a redundant check. 

Perhaps we could add the following error message to the old check ""Cannot import custom .glsl shaders when running without a RenderingDevice. This can happen if you are using the headless more or the Compatibility backend."

File: modules/betsy/image_compress_betsy.cpp [hunk 2]
```diff
@@ -36,11 +36,16 @@
 
 #include "bc1.glsl.gen.h"
 #include "bc6h.glsl.gen.h"
+#include "servers/display_server.h"
 
 static Mutex betsy_mutex;
 static BetsyCompressor *betsy = nullptr;
 
 void BetsyCompressor::_init() {
+	if (!DisplayServer::can_create_rendering_device()) {
```
Comment (clayjohn) [inline modules/betsy/image_compress_betsy.cpp:45]: There is still a risk of crashing here since Betsy might still try to run even if it returns here. 

We should also add an early return if `compress_rd == nullptr` in `_compress()`. It should return an Error so that the compression can fall back to CPU compression

**Review by clayjohn — COMMENTED** [review]
This looks mostly good. Just a bit more feedback to get it fully finished

File: modules/betsy/image_compress_betsy.cpp (hunk 2 above)
Comment (TCROC) [reply modules/betsy/image_compress_betsy.cpp:45]: done

File: servers/rendering/rendering_device_binds.cpp (hunk 1 above)
Comment (TCROC) [reply servers/rendering/rendering_device_binds.cpp]: done

**Comment by clayjohn** [comment]
Looks great! The final step before merging is to squash all the commits together so that the whole PR only contains 1 big commit with all your changes. We like to merge one commit at a time to keep the git history clean and navigable. 

If you don't know how to do that, we have a helpful tutorial in the official documentation https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase

**Comment by TCROC** [comment]
> Looks great! The final step before merging is to squash all the commits together so that the whole PR only contains 1 big commit with all your changes. We like to merge one commit at a time to keep the git history clean and navigable.
> 
> If you don't know how to do that, we have a helpful tutorial in the official documentation https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase

done :) Let me know if you need anything else :)

**Review by clayjohn — APPROVED** [review]
Looks great! Thank you

**Comment by TCROC** [comment]
> Looks great! Thank you

You are very welcome!  And thank you as well! ❤️

**Comment by Repiteo** [comment]
Thanks!
